Support UDTS for sequences and ISC-sequences view

aadityavermaa opened the pull request:
### Description
- Creation of view information_schema_tsql.sequences to support for T-SQL INFORMATION_SCHEMA.SEQUENCES.
- Provide support for creation of a sequence using user-defined data types which is currently not supported in Babelfish.
- Add test cases for the implementations

Pull request for changes in engine repository: 
https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/52

Task: BABEL-3504, BABEL-3452
Signed-off-by: Aditya Verma <adiityav@amazon.com>

### Testing 
#### Tests for ISC-sequences 
##### Use case based:
- Tests with all permissible data types
- Creating other objects from the view
- Tests for checking user-privilege
- Tests with sequences defined in different schemas
- Cross database tests 
##### Negative tests case scenarios
##### Arbitrary inputs 
##### Minor version upgrade tests
- Tests with all permissible data types
- Tests for checking user-privilege
- Tests with sequences defined in different schemas
- Cross database tests 
##### Major version upgrade tests
- same as use case based
#### Tests for BABEL-3504
##### Use case based:
- Test with default values
- Tests with all permissible data types
- Tests with UDTs of basetype as all permissible data types
- Tests for checking user-privilege
- Tests with UDT and sequence defined in different schema
##### Negative tests case scenarios:
- Test with non-permissible data type 
- Test with UDT of non-permissible basetype 
- Test of creating sequence with datatype that user does not have permission to use 
##### Arbitrary inputs 
##### Minor version upgrade tests

- ##### Major version upgrade tests
- same as use case based but psql test cases are for multi-db mode 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
BABEL-3504,3452:Support UDTS for sequences and ISC-sequences view

- Creation of view information_schema_tsql.sequences to support for T-SQL INFORMATION_SCHEMA.SEQUENCES.
- Provide support for creation of a sequence using user-defined data types which is currently not supported in Babelfish.
- Added support for cross-schema UDTs
- Add test cases for both implementations

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:Support UDTS for sequences and ISC-sequences view

- Creation of view information_schema_tsql.sequences to support for T-SQL INFORMATION_SCHEMA.SEQUENCES.
- Provide support for creation of a sequence using user-defined data types which is currently not supported in Babelfish.
- Add test cases for both implementations

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:Support UDTS for sequences and ISC-sequences view

- Creation of view information_schema_tsql.sequences to support for T-SQL INFORMATION_SCHEMA.SEQUENCES.
- Provide support for creation of a sequence using user-defined data types which is currently not supported in Babelfish.
- Added support for cross-schema
- Add test cases for both implementations

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:Support UDTS for sequences and ISC-sequences view

- Creation of view information_schema_tsql.sequences to support for T-SQL INFORMATION_SCHEMA.SEQUENCES.
- Provide support for creation of a sequence using user-defined data types which is currently not supported in Babelfish.
- Add test cases for both implementations

Signed-off-by: Aditya Verma <adiityav@amazon.com>

shalinilohia50 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1820,6 +1821,9 @@ pltsql_sequence_datatype_map(ParseState *pstate,
 		if (!for_identity || typmod_p != -1)
 		{
+			if (typmod_p == -1)
+                typmod_p = 1179652;
```
Nit: Indentation

jusjosj committed:
Support UDF as column default in ALTER TABLE (#581)

This change allows use of the syntax like

ALTER TABLE t2 ADD DEFAULT dbo.myf() FOR b

Previously, this type of statement would result in an error since we were not handling schema replacements in multidb.c for AT_ColumnDefault This PR adds logic to handle AT_ColumnDefault and calls the correct function to properly rewrite the dbo schema reference to its internal babelfish counter part

Task:BABEL-3268
Signed-off-by: Justin Jossick jusjosj@amazon.com

jusjosj committed:
Support UDF as column default in ALTER TABLE (#581)

This change allows use of the syntax like

ALTER TABLE t2 ADD DEFAULT dbo.myf() FOR b

Previously, this type of statement would result in an error since we were not handling schema replacements in multidb.c for AT_ColumnDefault This PR adds logic to handle AT_ColumnDefault and calls the correct function to properly rewrite the dbo schema reference to its internal babelfish counter part

Task:BABEL-3268
Signed-off-by: Justin Jossick jusjosj@amazon.com

jusjosj committed:
Support UDF as column default in ALTER TABLE (#581)

This change allows use of the syntax like

ALTER TABLE t2 ADD DEFAULT dbo.myf() FOR b

Previously, this type of statement would result in an error since we were not handling schema replacements in multidb.c for AT_ColumnDefault This PR adds logic to handle AT_ColumnDefault and calls the correct function to properly rewrite the dbo schema reference to its internal babelfish counter part

Task:BABEL-3268
Signed-off-by: Justin Jossick jusjosj@amazon.com

jusjosj committed:
Support UDF as column default in ALTER TABLE (#581)

This change allows use of the syntax like

ALTER TABLE t2 ADD DEFAULT dbo.myf() FOR b

Previously, this type of statement would result in an error since we were not handling schema replacements in multidb.c for AT_ColumnDefault This PR adds logic to handle AT_ColumnDefault and calls the correct function to properly rewrite the dbo schema reference to its internal babelfish counter part

Task:BABEL-3268
Signed-off-by: Justin Jossick jusjosj@amazon.com

Commit:
BABEL-3504,3452:Support UDTS for sequences and ISC-sequences view

- Creation of view information_schema_tsql.sequences to support for T-SQL INFORMATION_SCHEMA.SEQUENCES.
- Provide support for creation of a sequence using user-defined data types which is currently not supported in Babelfish.
- Added support for cross-schema UDTs
Add test cases for both implementations

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Changing location of typemod if condition

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

updated upgrade test files

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

updated upgrade test files

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452

added upgrade test files in input folder and ignored them in
jdbc_schedule

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

empty commit to re-run tests

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

empty commit to re-run tests

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

empty commit to re-run tests

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

empty commit to re-run tests

Signed-off-by: Aditya Verma <adiityav@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1749,9 +1749,21 @@ pltsql_sequence_datatype_map(ParseState *pstate,
 		return;
 
 	type_def = defGetTypeName(as_type);
+	type_names = type_def->names;
+	switch (list_length(type_def->names))
+    {
+        case 2:
+            strVal(linitial(type_names)) = get_physical_schema_name(get_cur_db_name(),strVal(linitial(type_names)));
+            break;
+        case 3:
+            strVal(lsecond(type_names)) = get_physical_schema_name(strVal(linitial(type_names)),strVal(lsecond(type_names)));
+            break;
+    }
+	*newtypid = typenameTypeId(pstate, type_def);
```
Fix indentation.
Why was it working earlier without this change? Is it because earlier it wasn't getting user defined datatypes?

aadityavermaa replied:
Yes, earlier there were some checks on the namespace name ( 'sys' and 'pg_catalog' ), no other case for schema name arises if User defined data types are not being used.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1837,6 +1853,12 @@ pltsql_sequence_datatype_map(ParseState *pstate,
 				(errmsg("NUMERIC or DECIMAL type is cast to BIGINT")));
 	}
+
+	if (aclresult != ACLCHECK_OK)
+    	aclcheck_error_type(aclresult, *newtypid);
```
Why are we not throwing the error just after permission check on line 1766?
Fix indentation.

aadityavermaa replied:
I observed and figured that it is logically incorrect; I've made the necessary changes in the newer commits.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1817,6 +1829,10 @@ pltsql_sequence_datatype_map(ParseState *pstate,
 		 * so it gets its own check
 		 */
+
+		if (typmod_p == -1)
+				typmod_p = 1179652;
```
Fix indentation.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1837,6 +1853,12 @@ pltsql_sequence_datatype_map(ParseState *pstate,
+	if (aclresult != ACLCHECK_OK)
+    	aclcheck_error_type(aclresult, *newtypid);
+
+    if(getBaseType(*newtypid)!=INT2OID || getBaseType(*newtypid)!=INT4OID || getBaseType(*newtypid)!=INT8OID) *newtypid = getBaseType(*newtypid);
```
Fix indentation. It is good coding practice to maintain 80 char limit in one line.

HarshLunagariya reviewed (changes requested):
Please add template for test cases covered in description.

surendravishnoi commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -302,3 +302,5 @@ sys-all_parameters-dep
 BABEL-3556
 BABEL-3268
 BABEL-sp_helpdb
+ISC-sequences
```
how does this test minor version upgrade? please add it to 14_5 to latest

surendravishnoi reviewed (commented):
Please use PR test template. Also fix indentation at various places.

KushaalShroff commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -716,4 +716,36 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+            CAST(s.seqmax AS bigint) AS "MAXIMUM_VALUE",
+            CAST(s.seqincrement AS bigint) AS "INCREMENT",
+            CAST( CASE WHEN s.seqcycle = 't' THEN 1 ELSE 0 END AS int) AS "CYCLE_OPTION",
+            CAST(NULL AS sys.nvarchar(128)) AS "DECLARED_DATA_TYPE",
+            CAST(NULL AS int) AS "DECLARED_NUMERIC_PRECISION",
+            CAST(NULL AS int) AS "DECLARED_NUMERIC_SCALE"
```
We need to update these columns for sequences with UDTs

aadityavermaa replied:
We can't update these columns as of now; since we have updated newtypid with the basetype of the UDT, we don't have the oid of the UDT stored in any object.

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1749,8 +1750,24 @@ pltsql_sequence_datatype_map(ParseState *pstate,
 
 	type_def = defGetTypeName(as_type);
+	type_names = type_def->names;
+	switch (list_length(type_def->names))
```
Nit: Use type_names

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1749,8 +1750,24 @@ pltsql_sequence_datatype_map(ParseState *pstate,
+			strVal(linitial(type_names)) = get_physical_schema_name(get_cur_db_name(),strVal(linitial(type_names)));
+			break;
+		case 3:
+			strVal(lsecond(type_names)) = get_physical_schema_name(strVal(linitial(type_names)),strVal(lsecond(type_names)));
```
This case is for db.schema.name, right?
In that case List will look like db,schema,name (list length of 3)
We need to convert it to db_schema,name (list length of 2)

what you have done makes the list look like: db,db_schema,name (list length of 3)

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1817,6 +1834,10 @@ pltsql_sequence_datatype_map(ParseState *pstate,
 		 * so it gets its own check
 		 */
+
+		if (typmod_p == -1)
```
please add an appropriate comment for this change.
1179652 is for Numeric(18,0), is that expected behaviour?

aadityavermaa replied:
In case of UDT typmod_p is equal to -1, but for a create sequence query for_identity is false, therefore the code enters the if condition below and the precision and scale is then calculated using typmod_p, which are wrong since typmod==-1. Hence, I've added this condition. I'll add the relevant comment.

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1837,6 +1858,11 @@ pltsql_sequence_datatype_map(ParseState *pstate,
 				(errmsg("NUMERIC or DECIMAL type is cast to BIGINT")));
 	}
+
+	if( getBaseType(*newtypid)!=INT2OID ||
```
save getBaseType(*newtypid) in a variable, rather than looking up 4 times

aadityavermaa replied:
Done

Commit:
BABEL-3504,3452:

addressed PR review comments

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

- added separate tests for 14_5->latest
- addressed PR comments
Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

changed data type of columns of information_schema.sequences from int to
sys.sql_variant

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

KushaalShroff commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql:
```diff
@@ -3672,6 +3672,41 @@ RETURNS sys.NVARCHAR(4000)
+GRANT SELECT ON information_schema_tsql.SEQUENCES TO PUBLIC;
+
+-- Reset search_path to not affect any subsequent scripts
+SELECT set_config('search_path', trim(leading 'sys, ' from current_setting('search_path')), false);
```
Why do we need this change?

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql:
```diff
@@ -3672,6 +3672,41 @@ RETURNS sys.NVARCHAR(4000)
+
+GRANT SELECT ON information_schema_tsql.SEQUENCES TO PUBLIC;
+
+-- Reset search_path to not affect any subsequent scripts
+SELECT set_config('search_path', trim(leading 'sys, ' from current_setting('search_path')), false);
+
```
why are we resetting the search_path?

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1749,8 +1751,27 @@ pltsql_sequence_datatype_map(ParseState *pstate,
+			strVal(linitial(type_names)) = get_physical_schema_name(get_cur_db_name(),strVal(linitial(type_names)));
+			break;
+		case 3:
+			strVal(lsecond(type_names)) = get_physical_schema_name(strVal(linitial(type_names)),strVal(lsecond(type_names)));
+			linitial(type_names) = lsecond(type_names);
```
instead of this two line, can't we just directly assign physical schema name to first element rather than assign it to second element first and then assign second element to first.

aadityavermaa replied:
Yes, that would be a better option. I have applied the change.

Commit:
BABEL-3504,3452:

- addressed review comments

Signed-off-by: Aditya Verma <adiityav@amazon.com>

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1749,8 +1751,26 @@ pltsql_sequence_datatype_map(ParseState *pstate,
+	switch (list_length(type_names))
+	{
+		case 2:
+			strVal(linitial(type_names)) = get_physical_schema_name(get_cur_db_name(),strVal(linitial(type_names)));
```
is it OK to change type names inside type_def->names

aadityavermaa replied:
The modified value is not being used in any of the parent functions ( i.e. init_params, DefineSequence, and ProcessUtilitySlow), therefore I believe it would not create any issues elsewhere.

KushaalShroff replied:
Surendra has a point, I think we should not change the state of "type_def" since it is extracted from "as_type" argument. We should remain transparent with the caller. Any change in the caller function, in future, could have an impact due to this change. We can rather copy this list locally and use it for our purpose inside pltsql_sequence_datatype_map.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1749,8 +1751,26 @@ pltsql_sequence_datatype_map(ParseState *pstate,
+			strVal(linitial(type_names)) = get_physical_schema_name(get_cur_db_name(),strVal(linitial(type_names)));
+			break;
+		case 3:
+			linitial(type_names) = get_physical_schema_name(strVal(linitial(type_names)),strVal(lsecond(type_names)));
```
why not strVal on left hand side

aadityavermaa replied:
It should be there. I have made the change.

Commit:
BABEL-3504,3452:

fixed error addressed in PR review

Signed-off-by: Aditya Verma <adiityav@amazon.com>

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1755,8 +1757,26 @@ pltsql_sequence_datatype_map(ParseState *pstate,
+		case 3:
+			strVal(linitial(type_names)) = get_physical_schema_name(strVal(linitial(type_names)),strVal(lsecond(type_names)));
+			lsecond(type_names) = list_nth_cell(type_names,2);
+			list_delete_last(type_names);
```
Please add a comment why you had to add this

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1843,6 +1871,12 @@ pltsql_sequence_datatype_map(ParseState *pstate,
 				(errmsg("NUMERIC or DECIMAL type is cast to BIGINT")));
 	}
+
+	base_type = getBaseType(*newtypid);
```
Lets move this at the top (appropriately) so that we can use it at 1840 as well

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1843,6 +1871,12 @@ pltsql_sequence_datatype_map(ParseState *pstate,
+	base_type = getBaseType(*newtypid);
+	if( base_type!=INT2OID ||
+		base_type!=INT4OID ||
+		base_type!=INT8OID) *newtypid = base_type;
```
Lets add a comment here.

Commit:
BABEL-3504,3452:

addressed PR comments in pl_handler.c

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Commit:
BABEL-3504,3452:

- Fixed timeout error occuring in JDBC test (github actions)

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-3504,3452:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1755,14 +1758,41 @@ pltsql_sequence_datatype_map(ParseState *pstate,
+		case 2:
+			new_type_names = list_make2(type_names->elements[0].ptr_value, type_names->elements[1].ptr_value);
+			strVal(linitial(new_type_names)) = get_physical_schema_name(get_cur_db_name(),strVal(linitial(type_names)));
+			type_def->names = new_type_names;
```
Move this outside switch case

HarshLunagariya replied:
+1

Commit:
BABEL-3504,3452:

-addressed review comment

Signed-off-by: Aditya Verma <adiityav@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1755,14 +1758,42 @@ pltsql_sequence_datatype_map(ParseState *pstate,
+		case 3:
+			/* Changing three part name of data type to physcial schema name */
+			new_type_names = list_make2(type_names->elements[1].ptr_value, type_names->elements[2].ptr_value);
+			strVal(linitial(new_type_names)) = get_physical_schema_name(strVal(linitial(type_names)),strVal(lsecond(type_names)));
```
nit: fix spacing after ','  -> strVal(linitial(type_names)),strVal(lsecond(type_names))

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1755,14 +1758,42 @@ pltsql_sequence_datatype_map(ParseState *pstate,
+	{
+		case 2:
+			new_type_names = list_make2(type_names->elements[0].ptr_value, type_names->elements[1].ptr_value);
+			strVal(linitial(new_type_names)) = get_physical_schema_name(get_cur_db_name(),strVal(linitial(type_names)));
```
nit: fix spacing after ',' -> strVal(get_cur_db_name(),strVal(lsecond(type_names))

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -1755,14 +1758,42 @@ pltsql_sequence_datatype_map(ParseState *pstate,
 	typ = typenameType(pstate, type_def, &typmod_p);
 	typname = typeTypeName(typ);
+	type_def->names = type_names;
+	if(list_len>1) list_free(new_type_names);
```
nit: fix spacing in `if(list_len>1) list_free(new_type_names);`
Should be
```
if (list_len > 1)
    list_free(new_type_names);
```

Commit:
BABEL-3504,3452:

- addressed review comments

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
fixed mixed declarations of some variables

Signed-off-by: Aditya Verma <adiityav@amazon.com>

surendravishnoi commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql:
```diff
@@ -7323,6 +7323,38 @@ RETURNS sys.NVARCHAR(4000)
 AS 'babelfishpg_tsql', 'tsql_query_to_json_text'
 LANGUAGE C IMMUTABLE COST 100;
 
+CREATE OR REPLACE VIEW information_schema_tsql.SEQUENCES AS
```
For ocelot, should be part of 2.3 to 2.4, not 2.2 to 2.3
